src/server/luaentity_sao.cpp

```diff
@@ -360,9 +360,17 @@ u16 LuaEntitySAO::punch(v3f dir,
 			&tool_item,
 			time_from_last_punch);
 
+	ItemStack old_selected_item;
+	puncher->getWieldedItem(&old_selected_item, nullptr);
+
 	bool damage_handled = m_env->getScriptIface()->luaentity_Punch(m_id, puncher,
 			time_from_last_punch, toolcap, dir, result.did_punch ? result.damage : 0);
 
+	puncher->getWieldedItem(&selected_item, nullptr);
+	// If LUA script added something to the selected slot launch item_OnWield
+	if (old_selected_item != selected_item)
+		puncher->getEnv()->getScriptIface()->item_OnWield(selected_item, puncher);
```

Comment:
Even if not very useful in practice, the convention is that Lua modifications to inventories does <i>not</i> run callbacks.

(This check also isn't very reliable if it only works in on_punch)